rufs/src/ufs/mod.rs

```diff
@@ -0,0 +1,148 @@
+use std::{
+	ffi::{OsStr, OsString},
+	fs::File,
+	io::{Cursor, Error as IoError, ErrorKind, Read, Result as IoResult, Seek, SeekFrom},
+	mem::size_of,
+	num::NonZeroU64,
+	os::unix::ffi::{OsStrExt, OsStringExt},
+	path::Path,
+};
+
+mod dir;
+mod inode;
+mod symlink;
+mod xattr;
+
+use crate::{
+	blockreader::BlockReader,
+	data::*,
+	decoder::{Config, Decoder},
+};
+
+#[macro_export]
+macro_rules! err {
+	($name:ident) => {
+		IoError::from_raw_os_error(libc::$name)
+	};
+}
+
+macro_rules! iobail {
+	($kind:expr, $($tk:tt)+) => {
+		return Err(IoError::new($kind, format!($($tk)+)))
+	};
+}
+
+#[derive(Debug, Clone)]
+pub struct Info {
+	pub blocks: u64,
+	pub bfree:  u64,
+	pub files:  u64,
+	pub ffree:  u64,
+	pub bsize:  u32,
+	pub fsize:  u32,
+}
+
+pub struct Ufs<R: Read + Seek> {
+	file:       Decoder<BlockReader<R>>,
+	superblock: Superblock,
+}
+
+impl Ufs<File> {
+	pub fn open(path: &Path) -> IoResult<Self> {
+		let file = BlockReader::open(path)?;
+		Self::new(file)
+	}
+}
+
+impl<R: Read + Seek> Ufs<R> {
+	pub fn new(mut file: BlockReader<R>) -> IoResult<Self> {
+		let pos = SBLOCK_UFS2 as u64 + MAGIC_OFFSET;
+		file.seek(SeekFrom::Start(pos))?;
+		let mut magic = [0u8; 4];
+		file.read_exact(&mut magic)?;
+
+		// magic: 0x19 54 01 19
+		#[cfg(not(fuzzing))]
+		let config = match magic {
+			[0x19, 0x01, 0x54, 0x19] => Config::little(),
+			[0x19, 0x54, 0x01, 0x19] => Config::big(),
+			_ => {
+				iobail!(
+					ErrorKind::InvalidInput,
+					"invalid superblock magic number: {magic:?}"
+				)
+			}
+		};
+		// FIXME: Choose based on hash of input or so, to excercise BE as well with introducing non-determinism
```

Comment:
I think that when fuzzing, you should guess the endianess just like you do when not fuzzing.  You can generate multiple corpuses, right?  One based on BE and one based on LE?